<a href="https://colab.research.google.com/github/hvasquez28/g1-ds_santander-f2-team7-Phyton/blob/main/Entrega_Final_Procesamiento_DanielTapia_LuisZanabria_NorbertoPena_HaroldVasquez/ProyectoPhytonGrupo7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <font color=green>Tabla de contenido</font>

[1. Identificación del problema](#problema)

[2. Planteamiento de Preguntas](#preguntas)

[3. Colección de Datos](#datos)

[4. Análisis Exploratorio de nuestro Dataset](#analisis)

[5. Limpieza de datos y agregaciones](#limpieza)

[6. Automatización y APIs](#API)

[7. Transformación, filtración y ordenamiento de datos](#ordenar)

[8. Planes a futuro](#futuro)

<a id="problema"></a>
# <font color=green>Identificación del Problema</font>

Para cualquier nuevo emprendedor, existe una serie de retos que lo abruman en su aventura de iniciar un nuevo negocio. La mayor incertidumbre es poder disponer de la información necesaria que lo apoye en la toma correcta de decisiones y así mitigar los riesgos de una mala inversión de tiempo y dinero.

Para esto, sería de gran ayuda proveerle a cada emprendedor de herramientas que le aporten información ordenada, certera y fidedigna; de tal forma que le otorgue seguridad y mayor certeza de si su idea será rentable o no.

Por lo tanto, sin lugar a duda, la **ciencia de datos** es la mejor herramienta que permitiría la creación de este tipo de herramientas de apoyo para un emprendedor y su nuevo negocio.

## Investigación al respecto del problema

En esta sección ira la investigación

## Búsqueda de soluciones anteriores

En esta sección irá las soluciones previas

<a id="preguntas"></a>
# <font color=green>Planteamiento de Preguntas </font>

De acuerdo con el problema planteado anteriormente, surgieron las siguientes preguntas tanto del emprendedor como del equipo:

1.  ¿Cuál es la mejor ubicación en Guadalajara, México para abrir mi negocio de bicicletas?
2. ¿Qué tantas ventas podré obtener en los primeros meses de iniciado mi negocio?
3. ¿Cuáles son los productos más demandados por los potenciales clientes?
4. ¿Qué precios serán los más competitivos para las bicicletas?
5. ¿Qué tan frecuente un cliente querrá comprar ropa o accesorios para sus bicicletas?
6. ¿Qué tantos clientes acudirán a mi negocio por una reparación o mejora en su bicicleta?

Aunque no tenemos seguridad si podremos responder todas las preguntas planteadas, se le comentó al emprendedor que se hará un análisis de la data disponible y se evaluará que información se le puede presentar.



<a id="datos"></a>
# <font color=green>Colección de Datos</font>
Se tomo la decision de usar "Bike Buyers 1000" y "Bike Sales", dado que contienen informacion relevante a la problematica sobre la cual se esta tratando y ambos datasets se complementan, los cuales fueron encontrados en kaggle.

"Bike Buyers 1000" 
link: "https://www.kaggle.com/datasets/heeraldedhia/bike-buyers"

"Bike Sales"
link: "https://www.kaggle.com/datasets/liyingiris90/bike-sales"

## Proceso para obtener el dataset

Dado que los archivos contenidos dentro del conjunto de datos de Bike Sales estan en formato .xlsx, se uso la funcion <font color =red> *pd.read_excel* </font> de pandas.

Para Bike Buyers al ser un formato csv, se usa nuestra funcion ya conocida  <font color =red> *pd.read_csv*</font>  

Se importan las siguientes librerias...

In [1]:
import pandas as pd
import numpy as np

Se guardan los respectivos DataFrames del conjunto de datos <font color=blue> "bike sales"</font> y <font color=blue> "bike buyers"

In [2]:
compradores = pd.read_csv("https://raw.githubusercontent.com/hvasquez28/g1-ds_santander-f2-team7-Phyton/Norberto/Entrega_Final_Procesamiento_DanielTapia_LuisZanabria_NorbertoPena_HaroldVasquez/datasets/bike_buyers.csv")

In [3]:
bikes = pd.read_excel('https://github.com/hvasquez28/g1-ds_santander-f2-team7-Phyton/blob/Norberto/Entrega_Final_Procesamiento_DanielTapia_LuisZanabria_NorbertoPena_HaroldVasquez/datasets/bikes.xlsx?raw=true')
bikestores = pd.read_excel("https://github.com/hvasquez28/g1-ds_santander-f2-team7-Phyton/blob/Norberto/Entrega_Final_Procesamiento_DanielTapia_LuisZanabria_NorbertoPena_HaroldVasquez/datasets/bikeshops.xlsx?raw=true")
orders = pd.read_excel("https://github.com/hvasquez28/g1-ds_santander-f2-team7-Phyton/blob/Norberto/Entrega_Final_Procesamiento_DanielTapia_LuisZanabria_NorbertoPena_HaroldVasquez/datasets/orders.xlsx?raw=true")

/usr/local/lib/python3.7/dist-packages/openpyxl/styles/stylesheet.py:226: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


Procedemos a revisar cada DataFrame usando <font color=blue> .head() </font> y <font color=blue>.tail()  </font>, asi como <font color=blue> .dtypes, .columns, .shape, .loc[$n:m$] </font> (donde $n,m ∈ 𝖭$ y $n<m.$) 

In [4]:
compradores.dtypes #Se tienen flotantes que deberian de ser enteros.
compradores.shape #Se tienen 1000 entradas y 13 columnas.
compradores.columns # ID, Marital S, Gender, Income, Children, Education, Occupation, Home Owner, Cars, Commute Distance, Region, Age, purchase bike
compradores.head() # Se observan NaNs
compradores.tail() # Se observan NaNs
c=list(compradores.columns) # Se guarda el nombre de las columnas en una lista, que sera usada mas adelante para cambiar los nombres de nuestro dataframe. 
compradores.loc[500:515] # Se observan Nans. 

ID Marital Status  Gender    Income  Children        Education  \
500  15559        Married    Male   60000.0       5.0        Bachelors   
501  19235        Married  Female   50000.0       0.0  Graduate Degree   
502  15275        Married    Male   40000.0       0.0  Partial College   
503  20339        Married  Female  130000.0       1.0        Bachelors   
504  25405        Married    Male   70000.0       2.0        Bachelors   
505  15940        Married    Male  100000.0       4.0  Partial College   
506  25074        Married  Female   70000.0       4.0        Bachelors   
507  24738        Married  Female   40000.0       1.0  Partial College   
508  16337        Married    Male   60000.0       0.0  Partial College   
509  24357        Married    Male       NaN       3.0        Bachelors   
510  18613         Single    Male   70000.0       0.0        Bachelors   
511  12207         Single    Male   80000.0       4.0        Bachelors   
512  18052        Married  Female   60000.0       1.0  Partial College   
513  13353         Single  Female   60000.0       4.0  Graduate Degree   
514  19399         Single    Male   40000.0       0.0        Bachelors   
515  16154        Married  Female   70000.0       5.0        Bachelors   

         Occupation Home Owner  Cars Commute Distance         Region   Age  \
500    Professional        Yes   1.0        2-5 Miles  North America  47.0   
501  Skilled Manual        Yes   0.0        0-1 Miles  North America  34.0   
502  Skilled Manual        Yes   1.0       5-10 Miles  North America  29.0   
503      Management        Yes   4.0        2-5 Miles  North America  44.0   
504  Skilled Manual        Yes   1.0        2-5 Miles  North America  38.0   
505    Professional        Yes   4.0        0-1 Miles  North America  40.0   
506    Professional        Yes   2.0        2-5 Miles  North America  42.0   
507        Clerical        Yes   1.0        1-2 Miles  North America  51.0   
508  Skilled Manual         No   2.0        1-2 Miles  North America  29.0   
509    Professional        Yes   1.0        2-5 Miles  North America  48.0   
510    Professional         No   1.0        2-5 Miles  North America  37.0   
511      Management        Yes   NaN       5-10 Miles  North America  66.0   
512  Skilled Manual        Yes   1.0        0-1 Miles  North America  45.0   
513      Management        Yes   2.0        10+ Miles  North America  61.0   
514    Professional         No   1.0        2-5 Miles  North America  45.0   
515    Professional        Yes   2.0        2-5 Miles  North America  47.0   

    Purchased Bike  
500             No  
501             No  
502             No  
503            Yes  
504            Yes  
505             No  
506            Yes  
507            Yes  
508             No  
509            Yes  
510            Yes  
511            Yes  
512            Yes  
513            Yes  
514             No  
515             No

In [5]:
bikes.dtypes # El unico data que parecer ser incorrecto es el precio, que es de tipo int. 
bikes.shape # Se tienen un total de 97 filas con 5 columnas
bikes.columns # bike.id, model, category1, category2, frame, price. Se renombraran columnas.
bikes.head() #No se observan Nans.
bikes.tail() #No se observan Nans.
bc=list(bikes.columns)
bikes.loc[46:58] # No se observan Nans de forma aleatoria

bike.id                  model category1           category2   frame  \
46       47        Jekyll Carbon 1  Mountain       Over Mountain  Carbon   
47       48        Jekyll Carbon 2  Mountain       Over Mountain  Carbon   
48       49        Jekyll Carbon 3  Mountain       Over Mountain  Carbon   
49       50        Jekyll Carbon 4  Mountain       Over Mountain  Carbon   
50       51       Trigger Carbon 1  Mountain       Over Mountain  Carbon   
51       52       Trigger Carbon 2  Mountain       Over Mountain  Carbon   
52       53       Trigger Carbon 3  Mountain       Over Mountain  Carbon   
53       54       Trigger Carbon 4  Mountain       Over Mountain  Carbon   
54       55  Scalpel-Si Black Inc.  Mountain  Cross Country Race  Carbon   
55       56        Scalpel-Si Race  Mountain  Cross Country Race  Carbon   
56       57    Scalpel-Si Hi-Mod 1  Mountain  Cross Country Race  Carbon   
57       58    Scalpel-Si Carbon 2  Mountain  Cross Country Race  Carbon   
58       59    Scalpel-Si Carbon 3  Mountain  Cross Country Race  Carbon   

    price  
46   7990  
47   6070  
48   3840  
49   3200  
50   8200  
51   5970  
52   3730  
53   3200  
54  12790  
55   9060  
56   7460  
57   6390  
58   5330

In [6]:
bikestores.dtypes # Todos los datos tienen el tipo correcto.
bikestores.shape # Se tienen un total de 30 filas con 6 columnas
bikestores.columns # bikeshop.id , bikeshop.name, bikeshop.city, bikeshop.state, latitude , longitude. Se renombraran columnas.
bikestores.head() # No se observan Nans.
bikestores.tail() # No se observan Nans.
bs=list(bikestores.columns)
bikestores.loc[10:25] #No se observan Nans. 

bikeshop.id                 bikeshop.name  bikeshop.city bikeshop.state  \
10           11               New York Cycles       New York             NY   
11           12                 Dallas Cycles         Dallas             TX   
12           13  Oklahoma City Race Equipment  Oklahoma City             OK   
13           14            Albuquerque Cycles    Albuquerque             NM   
14           15               Austin Cruisers         Austin             TX   
15           16               Phoenix Bi-peds        Phoenix             AZ   
16           17              Las Vegas Cycles      Las Vegas             NV   
17           18            Los Angeles Cycles    Los Angeles             CA   
18           19        San Francisco Cruisers  San Francisco             CA   
19           20              Portland Bi-peds       Portland             OR   
20           21        Seattle Race Equipment        Seattle             WA   
21           22               Ann Arbor Speed      Ann Arbor             MI   
22           23        Philadelphia Bike Shop   Philadelphia             PA   
23           24            Providence Bi-peds     Providence             RI   
24           25       New Orleans Velocipedes    New Orleans             LA   
25           26          Miami Race Equipment          Miami             FL   

     latitude   longitude  
10  40.712784  -74.005941  
11  32.776664  -96.796988  
12  35.467560  -97.516428  
13  35.085334 -106.605553  
14  30.267153  -97.743061  
15  33.448377 -112.074037  
16  36.169941 -115.139830  
17  34.052234 -118.243685  
18  37.774929 -122.419416  
19  45.523062 -122.676482  
20  47.606209 -122.332071  
21  42.280826  -83.743038  
22  39.952584  -75.165222  
23  41.823989  -71.412834  
24  29.951066  -90.071532  
25  25.761680  -80.191790

In [7]:
orders.dtypes # Se observa que order.id, order.line, customer.id y product.id son de tipo flotante cuando deberian de ser de tipo entero.
orders.shape # Tiene un total de 15644 filas y 7 columnas. 
orders.columns #'Unnamed: 0', 'order.id', 'order.line', 'order.date', 'customer.id','product.id' y 'quantity'. Se renombraran las columnas.
oc=list(orders.columns) 
orders.head() #No se observan Nans, por otro lado, se observa que la columna 'Unnamed: 0' esta repetida. 
orders.tail() # No se observan Nans. 
orders.loc[10468:10480] # no se observan Nans.

Unnamed: 0  order.id  order.line order.date  customer.id  product.id  \
10468       10469    1339.0         1.0 2014-07-08         25.0        12.0   
10469       10470    1339.0         2.0 2014-07-08         25.0        61.0   
10470       10471    1339.0         3.0 2014-07-08         25.0        72.0   
10471       10472    1339.0         4.0 2014-07-08         25.0        17.0   
10472       10473    1339.0         5.0 2014-07-08         25.0        42.0   
10473       10474    1339.0         6.0 2014-07-08         25.0        91.0   
10474       10475    1339.0         7.0 2014-07-08         25.0        50.0   
10475       10476    1339.0         8.0 2014-07-08         25.0        28.0   
10476       10477    1339.0         9.0 2014-07-08         25.0        39.0   
10477       10478    1339.0        10.0 2014-07-08         25.0        87.0   
10478       10479    1339.0        11.0 2014-07-08         25.0        88.0   
10479       10480    1339.0        12.0 2014-07-08         25.0        77.0   
10480       10481    1339.0        13.0 2014-07-08         25.0        23.0   

       quantity  
10468       1.0  
10469       1.0  
10470       1.0  
10471       1.0  
10472       2.0  
10473       1.0  
10474       1.0  
10475       1.0  
10476       1.0  
10477       8.0  
10478       1.0  
10479       1.0  
10480       1.0

Se crean diccionarios para cambiar el nombre de las columnas de nuestros data frames.

In [8]:
comp_names = {c[0]:'id_comprador',
              c[1]:'estado_civil',
              c[2]:'genero',
              c[3]:'sueldo',
              c[4]:'hijos',
              c[5]:'educacion',
              c[6]:"profesion",
              c[7]:"posee_casa",
              c[8]:'autos',
              c[9]:'distancia_d_viaje',
              c[10]:'region',
              c[11]:'edad',
              c[12]:'bicicleta_adquirida'}

bike_names = {bc[0]:'id_bicicleta',
              bc[1]:'modelo',
              bc[2]:'categoria_1',
              bc[3]:'categoria_2',
              bc[4]:'aleacion',
              bc[5]:'precio'}

stores_nombres = {bs[0]:'id_tienda',
                  bs[1]:'nombre_tienda',
                  bs[2]:'ciudad',
                  bs[3]:'estado',
                  bs[4]:'latitud',
                  bs[5]:'longitud'}

orders_nombres = {oc[1]:'id_orden',
                  oc[2]:'orden_line',
                  oc[3]:'orden_date',
                  oc[4]:'id_tienda',
                  oc[5]:'id_bicicleta',
                  oc[6]:'numero_items'}

Nos percatamos del hecho, de que la columna <font color=blue> customer_id</font>  es realmente el <font color=blue> id_tienda.</font> Dado que ambas coinciden con el numero de entradas y los valores asociados a cada una. 

In [9]:
print(orders['customer.id'].unique().max() )# Customer id es realmente bikeshop_id
print(orders['customer.id'].unique().min() ) # Minimo en la lista.
print(len(orders['customer.id'].unique()) ) # Tamaño de la lista.

30.0
1.0
30


#Renombramos las columnas en los DataFrames.

In [10]:
compradores = compradores.rename(columns=comp_names)
ordenes = orders.rename(columns=orders_nombres)
bicicletas = bikes.rename(columns=bike_names)
tiendas =  bikestores.rename(columns=stores_nombres)

#Se analizan los datos unicos de cada columna y se omiten las redundantes.

DataFrame ~  <font color=green> Compradores </font>

In [ ]:
print(list(compradores['coches'].unique())) #0,1,2,3,4
print(list(compradores['distancia_d_viaje'].unique())) #0-1, 1-2, 2-5, 5-10, 10+ 
print(list(compradores['edad'].unique())) # Se tiene un intervalo de edad de [25,89]
print(compradores['edad'].max()) #min 25 max 89 
print( len(list(compradores['educacion'].unique())))# 5 nivels de estudio. Donde el mas bajo no termino la prepa y el mas alto tiene un titulo. #'Bachelors', 'Partial College', 'High School', 'Partial High School', 'Graduate Degree'
print(len( list(compradores['profesion'].unique())))# 5 profesiones #'Skilled Manual', 'Clerical', 'Professional', 'Manual', 'Management'
print(list(compradores['region'].unique())) #'Europe', 'Pacific', 'North America'
print(list(compradores['sueldo'].unique())) # Se tienen sueldos dentro del rango [10000,170000]
print(compradores['sueldo'].min()) # max 170,000, min 10,000 

DataFrame ~  <font color=green> Ordenes </font> 

In [ ]:
print(list( ordenes['numero_items'].unique() )) # Se compran de 1 - 10 items. 
print(list( ordenes['orden_line'].unique() )) # Se realizan un maximo de 30 ordenes por cliente. [1,30] 

DataFrame ~  <font color=green> Bibicletas </font> 

In [ ]:
print(list( bicicletas['aleacion'].unique() )) #'Carbon', 'Aluminum'
print(list( bicicletas['categoria_1'].unique() )) # 'Road', 'Mountain'
print(len( list( bicicletas['categoria_2'].unique() ))) #Se tienen 9 categorias 2  # 'Elite Road', 'Endurance Road', 'Triathalon', 'Cyclocross', 'Over Mountain', 'Cross Country Race', 'Fat Bike', 'Trail', 'Sport'
print(len(list( bicicletas['modelo'].unique() ))) # Se tienen 97 modelos
print(list( bicicletas['precio'].unique() )) # El precio se encuentra dentro del rango [415,12790]
print(bicicletas['precio'].max()) #min 415 / max 12790 

DataFrame ~  <font color=green> Tiendas </font> 

In [ ]:
print(len( list( tiendas['ciudad'].unique() ))) # Se tiene un total de 30 ciudades  #'Pittsburgh', 'Ithaca', 'Columbus', 'Detroit', 'Cincinnati', 'Louisville', 'Nashville', 'Denver', 'Minneapolis', 'Kansas City', 'New York', 'Dallas', 'Oklahoma City', 'Albuquerque', 'Austin', 'Phoenix', 'Las Vegas', 'Los Angeles', 'San Francisco', 'Portland', 'Seattle', 'Ann Arbor', 'Philadelphia', 'Providence', 'New Orleans', 'Miami', 'San Antonio', 'Wichita', 'Indianapolis', 'Tampa'
print(len(list( tiendas['estado'].unique() )))# 21 estados # 'PA', 'NY', 'OH', 'MI', 'KY', 'TN', 'CO', 'MN', 'KS', 'TX', 'OK', 'NM', 'AZ', 'NV', 'CA', 'OR', 'WA', 'RI', 'LA', 'FL', 'IN'
print(len(list( tiendas['nombre_tienda'].unique() )))# y 30 tiendas. # 'Pittsburgh Mountain Machines', 'Ithaca Mountain Climbers', 'Columbus Race Equipment', 'Detroit Cycles', 'Cincinnati Speed', 'Louisville Race Equipment', 'Nashville Cruisers', 'Denver Bike Shop', 'Minneapolis Bike Shop', 'Kansas City 29ers', 'New York Cycles', 'Dallas Cycles', 'Oklahoma City Race Equipment', 'Albuquerque Cycles', 'Austin Cruisers', 'Phoenix Bi-peds', 'Las Vegas Cycles', 'Los Angeles Cycles', 'San Francisco Cruisers', 'Portland Bi-peds', 'Seattle Race Equipment', 'Ann Arbor Speed', 'Philadelphia Bike Shop', 'Providence Bi-peds', 'New Orleans Velocipedes', 'Miami Race Equipment', 'San Antonio Bike Shop', 'Wichita Speed', 'Indianapolis Velocipedes', 'Tampa 29ers'

<a id="analisis"></a>
# <font color=green>Análisis Exploratorio de nuestro Dataset</font>

En esta sección se busco responder las siguientes preguntas:

1. ¿El conjunto de datos que tengo realmente me sirve para responder algunas de las preguntas que me planteé?

> Luego de un primer analisis, se encontró que los <font color=purple>  DataSets</font> si pueden responder algunas de las preguntas planteadas. Sobre todo aquellas que se refieren economicamente. Es decir, se puede responder el numero de ventas iniciales por mes, el precio mas vendido, asi como los modelos mas demandados en bicletas, y sus caracteristicas que las componen.

2. ¿Qué tamaño tiene mi conjunto de datos? ¿Serán datos suficientes?

> El tamaño de mis conjunto de datos <font color=green> "bike sales" </font> es lo suficientemente grande para obtener informacion importante sobre los compradores, y sus preferencias al momento de comprar un bicicletas. Dado que esta compuesto por 15644 ordenes, que componen alrededor de 4 años. 

>Por otra parte se tienen diferentes tipos de bicletas que dan un total de 97. Aparte, hay 30 tiendas distribuidas en diferentes regiones del mundo y se tienen 1000 entradas respecto a clientes. Aun que, no todos ellos tomaron la decision de comprar bicicletas.

3. ¿Qué columnas tengo y qué información tengo en cada una de esas columnas?

> Se tienen alrededor de 31 columnas juntando todos los <font color=purple> DataFrames. </font>

> Para los <font color=purple> compadores </font>compadores tenemos estado Civil (Soltero o Casado), genero(Masculino o Femenino), sueldo (Este se encuentra entre 10,000 y 170,000 mensuales), hijos (De 0 a 4), educacion (Se tienen 5 niveles de estudio, donde el mas bajo no termino la prepa y el mas alto cuenta con un titulo.), profesion (Se tiene 5 profesiones), posesion de casa(Si/No), autos(De 0 a 4), distancia de viaje (0-10 millas), region (Europa, Norte America y el Pacifico), edad (La edad oscila entre 25 y 89 años) y compra de bicleta (Si/No).

> Para <font color=purple> bicicletas </font> se tiene la id, Modelo (Se tienen 97 modelos), categoria 1 (Urbana o Montaña), categoria 2 (Se tiene 9 Categorias), aleacion(Carbon o Aluminio), precio(Osila entre los 415 y 12,790 dolares).

> Para las <font color=purple> tiendas </font> se tiene el id, nombre de la tienda, ciudad (Se tienen 30 ciudades), estado (21 estados), latitud y longitud. 

> Para las <font color=purple> ordenes </font> tenemos id, linea de orden, fecha de orden, id tienda, id bicicleta, cantidad comprada(Se han comprado hasta 10 bicicletas por orden). 

4. Los nombres que tienen mis columnas, ¿son el nombre más apropiado?
> Inicialmente los datos vienen en ingles, ademas, las columnas estan escritas usando un formato distinto a <font color=green> *snake_case*.</font> Por lo que se toma la decision de renombrar las columnas usando <font color=green> *snake_case*</font> y traduciendolas al español. 

5. ¿Qué tipos de datos tengo en cada columna? ¿Parecen ser el tipo correcto de datos? ¿O es un tipo de datos "incorrecto"?
> Algunas columnas presentan datos de tipo flotante cuando debiesen de ser de tipo int. Y viceversa. Por lo que se cambian los tipos de datos en cada DataFrame que lo necesita.

6. Si selecciono algunas filas al azar y las observo, ¿estoy obteniendo los datos que debería? ¿o hay datos que parecen estar "sucios" o "incorrectos"?
> Tras aplicar <font color=blue>.head() </font>, <font color=blue> .tail() </font>, <font color=blue> .loc[n,m]. </font> Se observa que el unico DataFrame que presenta NaNs es el de <font color=purple> compadores </font>. Posteriormente se usaron las funciones <font color=green>*numero_nans* </font> y <font color=green> *porcentaje_nans*</font> para verificar que DataFrames realmente contenian NaNs y cuantas de cada columna.

<a id="limpieza"></a>
# <font color=green> Limpieza de datos y agregaciones </font>

En esta sección hicimos una limpieza de nuestros dataSets. Para ello, se sigueron los siguientes pasos:

1. Encontrar los NaNs que contiene
> Se crean funciones para determinar NaNs entre los distintos DataFrames.
Tras realizar este metodo a nuestro primer conjunto de datos ("Ordenes", "Bicicletas" y "Tiendas") nos damos cuenta que este no presenta Nans. Mientras que nuestro segundo conjunto de datos ("Compradores") si presenta Nans.

2. ¿Qué se hicieron con los NaNs?

> Para nuestro DataFrame "Compradores", se tienen un total de 53 NaNs, que representa al 5.3% de los datos. Tras aplicar nuestra funcion *numero_nans* observamos que hay en hijos, autos, estado_civil, genero, sueldo, posee_casa y edad.

> Para las filas con Nans de las columnas hijos y coches, estas seran llenadas con 0s. Mientras que para posee_casa y estado civil, con No y Single respectivamente. El resto de las filas con Nans seran eliminadas del DataFrame.  

> Finalmente nos queda un data frame con el 97.7% de nuestro conjunto de datos inicial.
  
3. Reindexo del DataSet

>Se hace uso del metodo *.reset_index(drop=True)* dado que no se necesita el indice de nuestro DataFrame "compradores". Dado que se tiene la columna id_comprador que es independiente al indice del DataFrame. 

4. Agregaciones al DataFrame

Se crean las siguientes funciones para determinar NaNs entre los distintos DataFrames.

In [11]:
def numero_nans(dataframe):
    print(dataframe.isna().sum())

In [12]:
def porcentaje_nans(dataframe):
  print(dataframe.isna().sum()/len(dataframe)*100)

In [13]:
porcentaje_nans(ordenes)
porcentaje_nans(tiendas)
porcentaje_nans(bicicletas)
porcentaje_nans(compradores) #El unico DataFrame con Nans presentes, es el de Compradores. 

Unnamed: 0      0.0
id_orden        0.0
orden_line      0.0
orden_date      0.0
id_tienda       0.0
id_bicicleta    0.0
numero_items    0.0
dtype: float64
id_tienda        0.0
nombre_tienda    0.0
ciudad           0.0
estado           0.0
latitud          0.0
longitud         0.0
dtype: float64
id_bicicleta    0.0
modelo          0.0
categoria_1     0.0
categoria_2     0.0
aleacion        0.0
precio          0.0
dtype: float64
id_comprador           0.0
estado_civil           0.7
genero                 1.1
sueldo                 0.6
hijos                  0.8
educacion              0.0
profesion              0.0
posee_casa             0.4
autos                  0.9
distancia_d_viaje      0.0
region                 0.0
edad                   0.8
bicicleta_adquirida    0.0
dtype: float64


In [15]:
numero_nans(compradores)

id_comprador            0
estado_civil            7
genero                 11
sueldo                  6
hijos                   8
educacion               0
profesion               0
posee_casa              4
autos                   9
distancia_d_viaje       0
region                  0
edad                    8
bicicleta_adquirida     0
dtype: int64


Se llanan las Nans. 

In [16]:
compradores[['hijos','autos']] = compradores[['hijos','autos']].fillna(0)
compradores['estado_civil'] = compradores['estado_civil'].fillna('Single')
compradores['posee_casa'] = compradores['posee_casa'].fillna('No')
compradores =compradores.dropna(how='any')
#numero_nans(compradores)
compradores = compradores.reset_index(drop=True)  

In [ ]:
numero_nans(compradores)

id_comprador           0
estado_civil           0
genero                 0
sueldo                 0
hijos                  0
educacion              0
profesion              0
posee_casa             0
coches                 0
distancia_d_viaje      0
region                 0
edad                   0
bicicleta_adquirida    0
dtype: int64


#Se procede a cambiar el tipo de dato de cada DataFrame.

Para esto crearemos diccionarios con los nuevos tipos de datos y aplicaremos la funcion de pandas *.astype()*

In [17]:
print(ordenes.dtypes) # Tiene flotantes que deben de ser enteros. 
print(tiendas.dtypes) # Tiendas tiene todos los tipos de datos correctos.
print(compradores.dtypes) # Tiene flotantes que deben de ser enteros. 
print(bicicletas.dtypes) # Tiene un int que deberia de ser flotante. 

Unnamed: 0               int64
id_orden               float64
orden_line             float64
orden_date      datetime64[ns]
id_tienda              float64
id_bicicleta           float64
numero_items           float64
dtype: object
id_tienda          int64
nombre_tienda     object
ciudad            object
estado            object
latitud          float64
longitud         float64
dtype: object
id_comprador             int64
estado_civil            object
genero                  object
sueldo                 float64
hijos                  float64
educacion               object
profesion               object
posee_casa              object
autos                  float64
distancia_d_viaje       object
region                  object
edad                   float64
bicicleta_adquirida     object
dtype: object
id_bicicleta     int64
modelo          object
categoria_1     object
categoria_2     object
aleacion        object
precio           int64
dtype: object


In [18]:
c=list(compradores.columns)
oc=list(ordenes.columns)
nuevos_types_compra = {c[4]:int,
         c[8]:int,
         c[11]:int}

new_types_orders ={oc[1]:int,
                   oc[2]:int,
                   oc[4]:int,
                   oc[5]:int,
                   oc[6]:int}

In [19]:
ordenes = ordenes.astype(new_types_orders)
compradores = compradores.astype(nuevos_types_compra)
bicicletas['precio'] = bicicletas['precio'].astype(float) # Se usa este metodo, dado que solo hay que cambiar de tipo de dato a una variable.

##Procederemos a eliminar la columna 'Unnamed: 0' del DataFrame ordenes

In [20]:
ordenes

Unnamed: 0  id_orden  orden_line orden_date  id_tienda  id_bicicleta  \
0               1         1           1 2011-01-07          2            48   
1               2         1           2 2011-01-07          2            52   
2               3         2           1 2011-01-10         10            76   
3               4         2           2 2011-01-10         10            52   
4               5         3           1 2011-01-10          6             2   
...           ...       ...         ...        ...        ...           ...   
15639       15640      2000           4 2015-12-25         15            14   
15640       15641      2000           5 2015-12-25         15            90   
15641       15642      2000           6 2015-12-25         15            17   
15642       15643      2000           7 2015-12-25         15            70   
15643       15644      2000           8 2015-12-25         15            61   

       numero_items  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1  
...             ...  
15639             1  
15640             1  
15641             1  
15642             1  
15643             1  

[15644 rows x 7 columns]

In [24]:
ordenes.drop(columns='Unnamed: 0',inplace=True)

In [25]:
ordenes

id_orden  orden_line orden_date  id_tienda  id_bicicleta  numero_items
0             1           1 2011-01-07          2            48             1
1             1           2 2011-01-07          2            52             1
2             2           1 2011-01-10         10            76             1
3             2           2 2011-01-10         10            52             1
4             3           1 2011-01-10          6             2             1
...         ...         ...        ...        ...           ...           ...
15639      2000           4 2015-12-25         15            14             1
15640      2000           5 2015-12-25         15            90             1
15641      2000           6 2015-12-25         15            17             1
15642      2000           7 2015-12-25         15            70             1
15643      2000           8 2015-12-25         15            61             1

[15644 rows x 6 columns]

In [26]:
bicicletas

id_bicicleta                          modelo categoria_1 categoria_2  \
0              1         Supersix Evo Black Inc.        Road  Elite Road   
1              2        Supersix Evo Hi-Mod Team        Road  Elite Road   
2              3  Supersix Evo Hi-Mod Dura Ace 1        Road  Elite Road   
3              4  Supersix Evo Hi-Mod Dura Ace 2        Road  Elite Road   
4              5      Supersix Evo Hi-Mod Utegra        Road  Elite Road   
..           ...                             ...         ...         ...   
92            93                         Trail 5    Mountain       Sport   
93            94                      Catalyst 1    Mountain       Sport   
94            95                      Catalyst 2    Mountain       Sport   
95            96                      Catalyst 3    Mountain       Sport   
96            97                      Catalyst 4    Mountain       Sport   

    aleacion   precio  
0     Carbon  12790.0  
1     Carbon  10660.0  
2     Carbon   7990.0  
3     Carbon   5330.0  
4     Carbon   4260.0  
..       ...      ...  
92  Aluminum    815.0  
93  Aluminum    705.0  
94  Aluminum    585.0  
95  Aluminum    480.0  
96  Aluminum    415.0  

[97 rows x 6 columns]

In [23]:
compradores

id_comprador estado_civil  genero    sueldo  hijos        educacion  \
0           12496      Married  Female   40000.0      1        Bachelors   
1           24107      Married    Male   30000.0      3  Partial College   
2           14177      Married    Male   80000.0      5  Partial College   
3           25597       Single    Male   30000.0      0        Bachelors   
4           13507      Married  Female   10000.0      2  Partial College   
..            ...          ...     ...       ...    ...              ...   
972         13466      Married    Male   80000.0      5  Partial College   
973         23731      Married    Male   60000.0      2      High School   
974         28672       Single    Male   70000.0      4  Graduate Degree   
975         19664       Single    Male  100000.0      3        Bachelors   
976         12121       Single    Male   60000.0      3      High School   

          profesion posee_casa  autos distancia_d_viaje         region  edad  \
0    Skilled Manual        Yes      0         0-1 Miles         Europe    42   
1          Clerical        Yes      1         0-1 Miles         Europe    43   
2      Professional         No      2         2-5 Miles         Europe    60   
3          Clerical         No      0         0-1 Miles         Europe    36   
4            Manual        Yes      0         1-2 Miles         Europe    50   
..              ...        ...    ...               ...            ...   ...   
972    Professional        Yes      3         1-2 Miles  North America    46   
973    Professional        Yes      2         2-5 Miles  North America    54   
974    Professional        Yes      0         2-5 Miles  North America    35   
975      Management         No      3         1-2 Miles  North America    38   
976    Professional        Yes      2         10+ Miles  North America    53   

    bicicleta_adquirida  
0                    No  
1                    No  
2                    No  
3                   Yes  
4                    No  
..                  ...  
972                  No  
973                 Yes  
974                 Yes  
975                  No  
976                 Yes  

[977 rows x 13 columns]

#Realizaremos un merge con los DataFrames Ordenes y Bicicletas para obtener resultados interesantes.
Se usara .merge()

In [29]:

orders_and_bikes = pd.merge(ordenes,bicicletas, left_on='id_bicicleta',right_index=True).sort_index()
orders_and_bikes

id_bicicleta  id_orden  orden_line orden_date  id_tienda  \
0                48         1           1 2011-01-07          2   
1                52         1           2 2011-01-07          2   
2                76         2           1 2011-01-10         10   
3                52         2           2 2011-01-10         10   
4                 2         3           1 2011-01-10          6   
...             ...       ...         ...        ...        ...   
15639            14      2000           4 2015-12-25         15   
15640            90      2000           5 2015-12-25         15   
15641            17      2000           6 2015-12-25         15   
15642            70      2000           7 2015-12-25         15   
15643            61      2000           8 2015-12-25         15   

       id_bicicleta_x  numero_items  id_bicicleta_y  \
0                  48             1              49   
1                  52             1              53   
2                  76             1              77   
3                  52             1              53   
4                   2             1               3   
...               ...           ...             ...   
15639              14             1              15   
15640              90             1              91   
15641              17             1              18   
15642              70             1              71   
15643              61             1              62   

                               modelo categoria_1         categoria_2  \
0                     Jekyll Carbon 3    Mountain       Over Mountain   
1                    Trigger Carbon 3    Mountain       Over Mountain   
2                 Beast of the East 2    Mountain               Trail   
3                    Trigger Carbon 3    Mountain       Over Mountain   
4      Supersix Evo Hi-Mod Dura Ace 1        Road          Elite Road   
...                               ...         ...                 ...   
15639                  CAAD12 Ultegra        Road          Elite Road   
15640                         Trail 3    Mountain               Sport   
15641                       CAAD8 105        Road          Elite Road   
15642                          F-Si 1    Mountain  Cross Country Race   
15643          Scalpel 29 Carbon Race    Mountain  Cross Country Race   

       aleacion  precio  
0        Carbon  3840.0  
1        Carbon  3730.0  
2      Aluminum  2130.0  
3        Carbon  3730.0  
4        Carbon  7990.0  
...         ...     ...  
15639  Aluminum  2240.0  
15640  Aluminum  1080.0  
15641  Aluminum  1410.0  
15642  Aluminum  2340.0  
15643    Carbon  6390.0  

[15452 rows x 13 columns]

In [70]:
specs =orders_and_bikes.groupby(['id_bicicleta','precio','aleacion','categoria_1','categoria_2']).count().sort_values(by=['id_orden'])

##El 53% de ventas se compone de aleaciones de Carbono, mientras que el 45% son de aluminio. 

In [75]:
orders_and_bikes.groupby('aleacion').size()/len(ordenes)*100

aleacion
Aluminum    45.512657
Carbon      53.260036
dtype: float64

El genero que mas compro bicicletas fue el Masculino

In [100]:
compradores.groupby('genero').size()

genero
Female    486
Male      491
dtype: int64

Las bicicletas mas compradas fueron

In [106]:
orders_and_bikes.groupby(['id_bicicleta']).size()

id_bicicleta
1     126
2     130
3     116
4     124
5     126
     ... 
92    191
93    203
94    164
95    190
96    182
Length: 96, dtype: int64

In [97]:
precios =orders_and_bikes.groupby(['id_bicicleta']).count().sort_values(by=)
print(precios.head())
print(precios.tail())
#precios.sort_values(=['precio'], ascending=False)

TypeError: ignored

In [55]:
orders_and_bikes.groupby(['id_bicicleta','aleacion','precio']).size()

id_bicicleta  aleacion  precio 
1             Carbon    10660.0    126
2             Carbon    7990.0     130
3             Carbon    5330.0     116
4             Carbon    4260.0     124
5             Carbon    3940.0     126
                                  ... 
92            Aluminum  815.0      191
93            Aluminum  705.0      203
94            Aluminum  585.0      164
95            Aluminum  480.0      190
96            Aluminum  415.0      182
Length: 96, dtype: int64

In [60]:
orders_and_bikes.groupby(['id_bicicleta','aleacion','precio']).agg(['mean','min','max','count']).sort_values(by=['precio'])

id_orden                 orden_line          \
                                      mean min   max count       mean min max   
id_bicicleta aleacion precio                                                    
96           Aluminum 415.0     960.631868   5  1981   182   7.681319   1  26   
95           Aluminum 480.0     989.973684  16  1995   190   8.489474   1  30   
94           Aluminum 585.0    1093.884146  16  1979   164   8.591463   1  28   
93           Aluminum 705.0    1012.330049   7  1968   203   7.246305   1  29   
92           Aluminum 815.0     887.539267  23  1990   191   8.497382   1  28   
...                                    ...  ..   ...   ...        ...  ..  ..   
21           Carbon   9590.0   1012.034286  18  1992   175   8.462857   1  25   
1            Carbon   10660.0  1001.698413   3  1980   126   8.253968   1  27   
65           Carbon   11190.0   956.601810   7  1995   221   8.357466   1  28   
80           Carbon   12250.0   993.514451  27  1995   173   8.387283   1  29   
54           Carbon   12790.0   922.791878  29  1942   197   9.502538   1  29   

                                                       orden_date             \
                              count                          mean        min   
id_bicicleta aleacion precio                                                   
96           Aluminum 415.0     182 2013-07-30 18:11:52.087912192 2011-01-11   
95           Aluminum 480.0     190 2013-08-20 18:18:56.842105344 2011-01-25   
94           Aluminum 585.0     164 2013-11-27 16:40:58.536585472 2011-01-25   
93           Aluminum 705.0     203 2013-09-11 09:48:46.108374272 2011-01-12   
92           Aluminum 815.0     191 2013-05-26 00:15:04.712041984 2011-02-04   
...                             ...                           ...        ...   
21           Carbon   9590.0    175 2013-09-14 18:47:18.857142784 2011-01-28   
1            Carbon   10660.0   126 2013-08-31 05:20:00.000000000 2011-01-10   
65           Carbon   11190.0   221 2013-07-18 15:05:42.081447936 2011-01-12   
80           Carbon   12250.0   173 2013-08-21 20:40:13.872832256 2011-02-10   
54           Carbon   12790.0   197 2013-06-24 16:19:29.543147264 2011-02-11   

                               ... id_bicicleta_x       numero_items          \
                               ...            max count         mean min max   
id_bicicleta aleacion precio   ...                                             
96           Aluminum 415.0    ...             96   182     1.280220   1   8   
95           Aluminum 480.0    ...             95   190     1.294737   1   8   
94           Aluminum 585.0    ...             94   164     1.250000   1   6   
93           Aluminum 705.0    ...             93   203     1.290640   1  10   
92           Aluminum 815.0    ...             92   191     1.366492   1  10   
...                            ...            ...   ...          ...  ..  ..   
21           Carbon   9590.0   ...             21   175     1.251429   1   7   
1            Carbon   10660.0  ...              1   126     1.277778   1   8   
65           Carbon   11190.0  ...             65   221     1.325792   1   6   
80           Carbon   12250.0  ...             80   173     1.260116   1   7   
54           Carbon   12790.0  ...             54   197     1.309645   1   8   

                                    id_bicicleta_y                
                              count           mean min max count  
id_bicicleta aleacion precio                                      
96           Aluminum 415.0     182           97.0  97  97   182  
95           Aluminum 480.0     190           96.0  96  96   190  
94           Aluminum 585.0     164           95.0  95  95   164  
93           Aluminum 705.0     203           94.0  94  94   203  
92           Aluminum 815.0     191           93.0  93  93   191  
...                             ...            ...  ..  ..   ...  
21           Carbon   9590.0    175           22.0  22

In [49]:
orders_and_bikes.sort_values(osize)

KeyError: ignored

In [ ]:
orders_and_bikes.

#Agregaciones

In [ ]:
compradores.groupby(['bicicleta_adquirida','genero','']).size()

bicicleta_adquirida  genero
No                   Female    250
                     Male      258
Yes                  Female    236
                     Male      233
dtype: int64

In [ ]:
compradores.groupby(['genero','estado_civil'],dropna=False).size()

genero  estado_civil
Female  Married         108
        Single          130
        NaN               1
Male    Married         120
        Single          115
        NaN               4
NaN     Married           1
        Single            2
dtype: int64

<a id="API"></a>
# <font color=green> Automatización y APIs </font>

Dado que se encontró un API interesante para el problema que se pelanteó, se decició usar uno y aplicar los pasos requeridos en esta sección, asi como el mismo analisis que se aplicó en las secciones nateriores. Para esto, se siguieron los siguientes pasos:

1. Encontrar el API
2. Leer la documentación
3. Realizar algunas peticiones de prueba
4. Automatizar el proceso de construcción del dataset
5 Explorar y limpiar el dataset
6. Unir los DataSet

<a id="ordenar"></a>
# <font color=green>Transformación, filtración y ordenamiento de datos </font>

<a id="futuro"></a>
# <font color=green> Planes a futuro </font>

1. ¿Qué podemos hacer con los datos que tenemos?
2. ¿Qué nos gustaría analizar?
3. ¿Qué información está disponible en los datos?
4. ¿Cómo podriamos ayudar a entender mejor los datos usando gráficas y visualizaciones? 